share/h2o/mruby/acl.rb

```diff
@@ -0,0 +1,130 @@
+# Copyright (c) 2016 DeNA Co., Ltd., Ichito Nagata
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+#
+require "h2o.rb"
+
+module H2O
+
+  module ACL
+
+    def acl(&block)
+      if @acl_handler then
+        raise "acl can be called only once for each handler configuration"
+      end
+      @acl_handler = ACLHandler.new(&block)
+      H2O.add_handler_validator(proc {|handler|
+        if handler != @acl_handler
+          raise "acl configuration is ignored"
+        end
+      })
+      return @acl_handler
+    end
+
+    class ACLHandler
+
+      class ConditionalHandler
+        def initialize(handler, cond)
+          @handler = handler
+          @cond = cond
+        end
+
+        def satisfy?(env)
+           return @cond.nil? || MatchingBlock.new(env).instance_eval(&@cond)
+        end
+
+        def call(env)
+          return @handler.call(env) if satisfy?(env)
+          return [399, {}, []]
+        end
+      end
+
+      def initialize(&block)
+        @acl = []
+        @returned = instance_eval(&block)
```

Comment:
Doesn't this assignment cause the _last_ access-control directive to be registered to both `@acl` and `@returned`?

Generalizing the issue, it is unclear to me why we'd want to return values from each directive (e.g. `allow`, `deny`). If there's a need to do so, please explain why (and also create a test case covering such use-case). If there's no need, we should stop returning values from the directives and delete `@returned`.